### ISSUE TITLE
Improve CNN Model Accuracy and Reduce Loss with Mixed Image Formats (PNG, JPG, etc.)

### ISSUE THREAD

**Issue by TranHuuDat2004** [TranHuuDat2004/handwriting-signature-recognition#1]
# Title: Improve CNN Model Accuracy and Reduce Loss with Mixed Image Formats (PNG, JPG, etc.)

**Labels:** `enhancement`, `bug`, `data-processing`, `model-accuracy`

---

## 1. Description

This issue addresses the challenge of maintaining high accuracy and low loss in our handwriting signature writer identification CNN model when dealing with a dataset containing mixed image formats, specifically PNG and JPG.

Currently, the model demonstrates high performance when the entire dataset consists of PNG images. However, introducing JPG images (or potentially other formats like TIFF) significantly degrades performance, leading to incorrect predictions and, in some cases, model instability or failure during training/inference.

## 2. Current Behavior

*   The model achieves high accuracy (e.g., >95% on validation) when trained and tested exclusively on PNG images.
*   When JPG images are included in the dataset (either for training, validation, or testing):
    *   Prediction accuracy drops noticeably.
    *   Validation loss may increase or fluctuate erratically.
    *   The model may misclassify signatures it previously identified correctly.
    *   In severe cases, errors might occur during the image loading and preprocessing pipeline if not handled robustly for different formats.

## 3. Expected Behavior

*   The CNN model should be robust to common image formats like PNG, JPG, and ideally TIFF without a significant drop in accuracy or increase in loss.
*   The model should maintain a high level of accuracy (e.g., comparable to the PNG-only performance) when trained and evaluated on datasets with mixed image formats.
*   The image preprocessing pipeline should consistently handle different image formats, ensuring uniform input tensor characteristics (dimensions, channels, normalization) for the model.

## 4. Possible Causes / Areas to Investigate

*   **Image Decoding Differences:** `tf.image.decode_png` and `tf.image.decode_jpeg` might handle aspects like color channels, alpha channels, or compression artifacts differently, leading to subtle variations in the input tensors even after resizing and grayscale conversion.
*   **Compression Artifacts:** JPG is a lossy compression format. Artifacts introduced by JPG compression might be misinterpreted by the model as features, or they might obscure actual signature features.
*   **Channel Handling:** Ensuring consistent channel count (e.g., 1 for grayscale) across all decoded images is crucial. The current code has checks:
    ```python
    if image.shape[-1] != CHANNELS and CHANNELS == 1:
         image = tf.image.rgb_to_grayscale(image)
    elif image.shape[-1] != CHANNELS and CHANNELS == 3 and image.shape[-1] == 1:
         image = tf.image.grayscale_to_rgb(image)
    ```
    We need to verify if this is sufficient for all cases, especially for JPGs which might sometimes be read with an unexpected number of channels by default.
*   **Color Space Conversion:** Differences in how PNG (often RGBA or RGB) and JPG (often YCbCr internally, converted to RGB) are handled might introduce subtle color/intensity variations before grayscale conversion.
*   **Metadata/EXIF Data:** While less likely to directly affect pixel values after decoding, some image formats might carry metadata that could (in rare cases) influence decoding libraries.
*   **Data Augmentation:** If data augmentation is applied, its effects might vary slightly depending on the initial quality/characteristics of PNG vs. JPG images.

## 5. Proposed Solutions / Steps to Reproduce & Fix

*   **Standardize Image Loading and Conversion:**
    *   Ensure all images, regardless of original format, are explicitly converted to a standard format (e.g., RGB) *before* being converted to grayscale. This might involve using `tf.image.convert_image_dtype` or more explicit PIL/OpenCV handling if TensorFlow's default decoders are inconsistent.
    *   Consider always decoding to 3 channels first, then converting to grayscale, to have a more uniform pipeline:
      ```python
      # Example of a more robust decoding/conversion
      try:
          image = tf.image.decode_image(image_bytes, channels=3, expand_animations=False) # Decode to 3 channels
      except tf.errors.InvalidArgumentError:
          # Fallback or error
          print(f"Could not decode image {path} to 3 channels")
          return None, label
      if CHANNELS == 1:
          image = tf.image.rgb_to_grayscale(image)
      ```
*   **Investigate Pixel Value Ranges:** After decoding and before normalization, inspect and compare the typical pixel value distributions of PNG vs. JPG images to see if there are systematic differences.
*   **More Robust Preprocessing:**
    *   Experiment with image normalization techniques beyond simple `/ 255.0`.
    *   Consider applying a slight Gaussian blur to potentially smooth out JPG compression artifacts before feeding to the model, though this might also remove fine details.
*   **Data Augmentation Strategies:**
    *   Introduce augmentations that mimic compression artifacts (e.g., `tf.image.random_jpeg_quality`) during training, even for PNG images, to make the model more robust to such variations.
*   **Thorough Testing with Mixed Datasets:**
    *   Create specific small datasets with a mix of PNG and JPG versions of the *same* signatures (if possible) to isolate the impact of format.
    *   Systematically evaluate the model on training/validation sets with varying PNG/JPG ratios.
*   **Inspect Intermediate Tensors:** Visualize the preprocessed images (tensors fed into the model) for both PNG and JPG inputs to spot any visible differences that might be affecting the model.
*   **Error Analysis:** When the model fails on JPGs, analyze which specific images are causing problems and what their characteristics are.

## 6. Environment

*   **TensorFlow Version:** [e.g., 2.10.0 - as per your `print(f"TensorFlow version: {tf.__version__}")`]
*   **Python Version:** [e.g., 3.8]
*   **Operating System:** [e.g., Google Colab (Linux)]
*   Any other relevant libraries and their versions.

## 7. Additional Context (Optional)

*   The current preprocessing function `load_and_preprocess_image` and `preprocess_single_image_for_prediction` attempts to handle PNG and JPG, but issues arise when they are mixed.
*   The goal is a production-ready model that can reliably handle common user-uploaded image formats.

---

By addressing this issue, we aim to significantly enhance the model's robustness and practical applicability.
